src/movies.js

```diff
@@ -1,16 +1,199 @@
+//         "Crime",
+//         "Drama"
+//       ],
+//       "rate": 9
+//     },  {
+//         "title": "The Godfather: Part dsII",
+//         "year": 1974,
+//         "director": "Francis Ford Coppola",
+//         "duration": "3h",
+//         "genre": [
+//           "Crime",
+//           "Drama"
+//         ],
+//         "rate": 4
+//       },
+//     {
+//       "title": "The Dark Knight",
+//       "year": 2008,
+//       "director": "Christopher Nolan",
+//       "duration": "2h 32min",
+//       "genre": [
+//         "Action",
+//         "Crime",
+//         "Drama",
+//         "Thriller"
+//       ],
+//       "rate": 9
+//     }]
+    
+    // Iteration 1: All directors? - Get the array of all directors.
+ 
+
+// console.log(getAllDirectors(trials));
+function getAllDirectors(movies) {
+    const unfilteredDirs = movies.map(function(movie) {
+        return movie.director;
+    });
+    const filteredDirs = unfilteredDirs.filter(function (director, index) {
+        if (unfilteredDirs.indexOf(director, index + 1) === -1) return true;
+        else return false;
+    });
+    return filteredDirs;
+};
+
 // _Bonus_: It seems some of the directors had directed multiple movies so they will pop up multiple times in the array of directors. How could you "clean" a bit this array and make it unified (without duplicates)?
 
+
+
 // Iteration 2: Steven Spielberg. The best? - How many drama movies did STEVEN SPIELBERG direct?
+function howManyMovies(movies) {
+    return movies.filter(function (movie) {
+        if (movie.director === 'Steven Spielberg' && movie.genre.includes("Drama")) return true;
+    }).length;
+}
+
 
 // Iteration 3: All rates average - Get the average of all rates with 2 decimals
 
+function ratesAverage(movies) {
+    if (movies.length === 0) return 0;
+    const sum = movies.reduce(function (total, movie) {
+        movie.rate = movie.rate || 0;
+        return total + movie.rate;    
+    }, 0);
+
+    const avg = Number((sum / movies.length).toFixed(2));
+    return avg;
+}
+
+// console.log(ratesAverage([{ rate: 6 }, { rate: '' }, {}]));
+
+
 // Iteration 4: Drama movies - Get the average of Drama Movies
 
+function dramaMoviesRate(movies) {
+    const dramaMovies = movies.filter(function(movie) {
+        if (movie.genre.includes("Drama")) return true;
+    });
+    return ratesAverage(dramaMovies);
+}
+
+// console.log(dramaMoviesRate(trials));
+
+
 // Iteration 5: Ordering by year - Order by year, ascending (in growing order)
 
+function orderByYear(movies) {
+    return [...movies].sort(function (a,b) {
+        return a.year - b.year || a.title.localeCompare(b.title);
+    })
+}
+
+// console.log(orderByYear([
+//     { title: 'abc', year: 2002 },
+//     { title: 'bac', year: 1982 },
+//     { title: 'aab', year: 1982 }
+//   ]))
+
 // Iteration 6: Alphabetic Order - Order by title and print the first 20 titles
+function orderAlphabetically(movies){
+    return [...movies].sort( function (a, b) {
```

Comment:
A lot is happening on this line and the following lines.
I would either place each method on a new line or I would even store each operation in a variable.

`.splice()`
`.map()`

Or

`const firstTwentyMovies = ////.splice(0, 20)` etc.